File: 112-path-sum/cpp/main2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
  // iterative solution
 public:
  bool hasPathSum(TreeNode* root, int sum) {
    if (root == NULL) {
      return false;
    }

    stack<TreeNode*> s;
    stack<int> sumStack;

    s.push(root);
    sumStack.push(0);

    while (!s.empty()) {
      auto node = s.top();
      s.pop();
      auto sumSoFar = sumStack.top();
      sumStack.pop();

      if (node->left == NULL && node->right == NULL &&
          node->val + sumSoFar == sum) {
        return true;
      }

      if (node->left != NULL) {
        s.push(node->left);
        sumStack.push(node->val + sumSoFar);
      }

      if (node->right != NULL) {
        s.push(node->right);
        sumStack.push(node->val + sumSoFar);
      }
    }

    return false;
  }
};

File: 112-path-sum/go/main2.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func hasPathSum(root *TreeNode, sum int) bool {
	if root == nil {
		return false
	}

	nodeStack := []*TreeNode{}
	sumStack := []int{}

	nodeStack = append(nodeStack, root)
	sumStack = append(sumStack, 0)

	for len(nodeStack) != 0 {
		node := nodeStack[len(nodeStack)-1]
		nodeStack = nodeStack[:len(nodeStack)-1]

		sumSoFar := sumStack[len(sumStack)-1]
		sumStack = sumStack[:len(sumStack)-1]

		if node.Left == nil && node.Right == nil && sumSoFar+node.Val == sum {
			return true
		}

		if node.Left != nil {
			nodeStack = append(nodeStack, node.Left)
			sumStack = append(sumStack, sumSoFar+node.Val)
		}

		if node.Right != nil {
			nodeStack = append(nodeStack, node.Right)
			sumStack = append(sumStack, sumSoFar+node.Val)
		}
	}

	return false
}
